runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1174,23 +1029,400 @@ static iree_status_t iree_hal_hip_device_queue_write(
-      options));
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_device_queue_write_streaming(
+              base_device, queue_affinity, wait_semaphore_list,
+              signal_semaphore_list, source_buffer, source_offset, target_file,
+              target_offset, length, flags, options));
+  IREE_TRACE_ZONE_END(z0);
   return loop_status;
 }
 
-static void iree_hal_hip_device_collect_tracing_context(void* user_data) {
-  iree_hal_stream_tracing_context_collect(
-      (iree_hal_stream_tracing_context_t*)user_data);
+typedef struct iree_hal_hip_device_semaphore_submit_callback_data_t {
+  iree_atomic_ref_count_t wait_semaphore_count;
+  iree_hal_hip_device_t* device;
+  iree_hal_queue_affinity_t queue_affinity;
+  iree_hal_command_buffer_t* command_buffer;
+  iree_hal_buffer_binding_table_t binding_table;
+  iree_hal_semaphore_list_t wait_semaphore_list;
+  iree_hal_semaphore_list_t signal_semaphore_list;
+  iree_hal_resource_set_t* resource_set;
+  iree_slim_mutex_t status_mutex;
+  iree_status_t status;
+} iree_hal_hip_device_semaphore_submit_callback_data_t;
+
+static iree_status_t iree_hal_hip_device_complete_submission(
+    void* user_data, iree_hal_hip_event_t* event, iree_status_t status) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  iree_hal_hip_device_semaphore_submit_callback_data_t* data =
+      (iree_hal_hip_device_semaphore_submit_callback_data_t*)user_data;
+  iree_hal_hip_device_t* device = data->device;
+
+  // Get the device_context from the queue_affinity.
+  uint32_t idx = iree_math_count_trailing_zeros_u64(data->queue_affinity);
+
+  // 1) Read any tracing events that were submitted.
+
+  iree_hal_command_buffer_t* command_buffer = data->command_buffer;
+  if (iree_hal_hip_multi_queue_command_buffer_isa(command_buffer)) {
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_hip_multi_queue_command_buffer_get(
+                command_buffer, data->queue_affinity, &command_buffer));
+  }
+
+  if (iree_hal_hip_stream_command_buffer_isa(command_buffer)) {
+    status = iree_status_join(
+        status,
+        iree_hal_stream_tracing_context_collect_list(
+            // Get the tracing context from the device/stream/queue affinity.
+            device->topology.devices[idx].tracing_context,
+            // Get the tracing event list from the command buffer.
+            iree_hal_hip_stream_command_buffer_tracing_events(command_buffer)
+                .head));
+  } else if (iree_hal_hip_graph_command_buffer_isa(command_buffer)) {
+    status = iree_status_join(
+        status,
+        iree_hal_stream_tracing_context_collect_list(
+            // Get the tracing context from the device/stream/queue affinity.
+            device->topology.devices[idx].tracing_context,
+            // Get the tracing event list from the command buffer.
+            iree_hal_hip_graph_command_buffer_tracing_events(command_buffer)
+                .head));
+  }
+
+  // 2) Free the event we specifically created here.
```

Comment:
"here" is ambiguous (it wasn't created in this function, and since it's a callback it's unclear without tracing back to the origin when it was created)